File: backend/app/schemas/complaint.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


class ComplaintCreate(BaseModel):
    title: str = Field(min_length=3, max_length=150)
    description: str = Field(min_length=3, max_length=5000)
    category: str = Field(min_length=2, max_length=60)
    priority: str = Field(default="MEDIUM", pattern="^(LOW|MEDIUM|HIGH|CRITICAL)$")


class ComplaintUpdate(BaseModel):
    status: str | None = Field(default=None, pattern="^(OPEN|ASSIGNED|IN_PROGRESS|RESOLVED|CLOSED)$")
    assigned_to_user_id: int | None = None
    resolution: str | None = Field(default=None, max_length=5000)


class ComplaintEscalate(BaseModel):
    reason: str = Field(min_length=3, max_length=500)


class ComplaintAttachmentResponse(BaseModel):
    id: int
    file_name: str
    content_type: str
    file_size: int
    file_url: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ComplaintResponse(BaseModel):
    id: int
    organization_id: int
    resident_id: int
    title: str
    description: str
    category: str
    priority: str
    status: str
    assigned_to_user_id: int | None
    escalated_at: datetime | None
    escalation_reason: str | None
    resolution: str | None
    resolved_at: datetime | None
    created_by_user_id: int
    created_at: datetime
    updated_at: datetime
    attachments: list[ComplaintAttachmentResponse] = []
    model_config = ConfigDict(from_attributes=True)

File: frontend/safecolony_app/lib/features/complaints/models/complaint.dart
import '../../../core/utils/api_date_time.dart';

class ComplaintAttachment {
  final int id;
  final String fileName;
  final String contentType;
  final int fileSize;
  final String fileUrl;
  final DateTime createdAt;

  const ComplaintAttachment({
    required this.id,
    required this.fileName,
    required this.contentType,
    required this.fileSize,
    required this.fileUrl,
    required this.createdAt,
  });

  factory ComplaintAttachment.fromJson(Map<String, dynamic> j) => ComplaintAttachment(
        id: (j['id'] as num).toInt(),
        fileName: j['file_name']?.toString() ?? 'Attachment',
        contentType: j['content_type']?.toString() ?? '',
        fileSize: (j['file_size'] as num?)?.toInt() ?? 0,
        fileUrl: j['file_url']?.toString() ?? '',
        createdAt: ApiDateTime.parse(j['created_at']),
      );
}

class Complaint {
  final int id;
  final String title;
  final String description;
  final String category;
  final String priority;
  final String status;
  final int? assignedToUserId;
  final DateTime createdAt;
  final DateTime? resolvedAt;
  final String? resolution;
  final List<ComplaintAttachment> attachments;

  Complaint({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.priority,
    required this.status,
    this.assignedToUserId,
    required this.createdAt,
    this.resolvedAt,
    this.resolution, required this.attachments,
  });

  factory Complaint.fromJson(Map<String, dynamic> j) => Complaint(
    id: j['id'],
    title: j['title'] ?? '',
    description: j['description'] ?? '',
    category: j['category'] ?? '',
    priority: j['priority'] ?? 'MEDIUM',
    status: j['status'] ?? 'OPEN',
    assignedToUserId: j['assigned_to_user_id'],
    createdAt: ApiDateTime.parse(j['created_at']),
    resolvedAt: j['resolved_at'] == null ? null : ApiDateTime.parse(j['resolved_at']),
    resolution: j['resolution'],
    attachments: (j['attachments'] as List? ?? []).map((e) => ComplaintAttachment.fromJson(Map<String,dynamic>.from(e))).toList(),
  );
}
